### PR TITLE
Require file extensions for imports and exports

### LINKED ISSUE #3488: Type imports broken in TypeScript's NodeNext mode (ESM support)
### What’s the bug you are facing?

This used to work in `beta.199`:

```ts
import type { JSONContent } from "@tiptap/core";
```

Now (`beta.204`) it shows the `Module has no exported member` error.

This could be due to this [TS issue](https://github.com/microsoft/TypeScript/issues/50482), meaning the `index.d.ts` exported by tiptap is using expressions like `export * from './CommandManager';`, while it should be `export * from './CommandManager.js';`.

I'm using:
-  ESM, i.e. `"type": "module"` in my `package.json`.
- TS version `4.9.3`.
- TS `NodeNext` mode, as supposed to be used for ESM projects.

### Which browser was this experienced in? Are any special extensions installed?

This happens in TS.

### How can we reproduce the bug on our side?

1. Create an ESM TS project with `"type": "module"` in your `package.json` and `"module": "NodeNext"` in your `tsconfig.json`.
2. Write `import type { JSONContent } from "@tiptap/core";`
3. Run `tsc`
4. See the error.

### Can you provide a CodeSandbox?

No, that's in TS.

### What did you expect to happen?

TS should resolve the type.

### Anything to add? (optional)

_No response_

### Did you update your dependencies?

- [X] Yes, I’ve updated my dependencies to use the latest version of all packages.

### Are you sponsoring us?

- [ ] Yes, I’m a sponsor. 💖

### PR DESCRIPTION
**Pull request by benasher44**
## Please describe your changes

Updates all imports and exports to include file extensions.

## How did you accomplish your changes

Careful find/replace, eslint, and iterations of building and fixing issues.

## How have you tested your changes

Built the project and demos, tested the changes with our main project, and tried a few examples in the deployed demos.

## How can we verify your changes

I'm not really sure how you all test something like this, but it sounded like from a comment in #3488 there was some method in mind. Let me know how I can help here.

## Remarks

- Updated an eslint rule to ensure these changes will stick going forward
- Used no-restricted-imports to specifically target '..' and '.' imports, which weren't covered by `import/extensions` for some reason.
- Generally speaking, I think this should work fine with CJS and improve compatibility for TS folks migrating to ESM. I'm less familiar with other supported packaging/module models.
- In cases where Vite/Rollup didn't like .js extensions for .ts files when building demos, I used .ts endings.

## Checklist

- [x] The changes are not breaking the editor
- [x] Added tests where possible
- [x] Followed the guidelines
- [x] Fixed linting issues

## Related issues

Closes #3488

### CODE REVIEW

**Comment by benasher44** [comment]
In my latest update, I did a more comprehensive test build again our internal app and found a final issue that wasn't caught by the updated eslint rule. I filed it as a bug in import-js/eslint-plugin-import#2777 and added a no-restricted-imports rule to cover the imports from ".." and "." case

**Comment by benasher44** [comment]
Hi @bdbch! Any thoughts here? prosemirror also recently published nodenext-compatible updates as well, so with this, we should be all set in that area.

**Comment by benasher44** [comment]
Hi @bdbch. Is this solution okay, or were you looking for something different? Would love to get this merged. This is the only library we need to fix in order to migrate our node backend to ESM.

**Comment by bdbch** [comment]
Hey @benasher44, sorry for the late update. I'll give your PR a try later this day and will merge if it works! Wanted to do this too earlier to bring in support for NodeNext.

Thanks already for contributing and keeping me in the loop

**Comment by benasher44** [comment]
Amazing, thank you! I have updated develop. I'll watch CI to make sure it passes.

**Comment by benasher44** [comment]
It looks like 1 workflow still has to be approved to run

**Comment by benasher44** [comment]
Okay I added an extensionAlias to the webpack ts-loader config that cypress is using, so it should properly understand/resolve the .js endings now. I also updated with develop again. Mind approving the workflow run once more? 🤞

**Comment by bdbch** [comment]
Thanks @benasher44! Let's see what the CI is doing - afterwards the PR looks good for me. I'll let @svenadlung also take a look.

File: .eslintrc.js
Comment (bdbch) [inline .eslintrc.js:46]: Do you know if this is autofixeable?

**Review by bdbch — APPROVED** [review]

File: .eslintrc.js
Comment (benasher44) [reply .eslintrc.js:46]: I don't think it is. The only way to get something like that is to convert all of the packages to use moduleResolution: NodeNext, and then TypeScript will suggest imports with .js endings from the start.

I can try that, if you all are open to it. It's a bigger change / commitment.

Comment (benasher44) [reply .eslintrc.js:46]: I briefly tried locally, and it creates a bunch of TS type resolution issues that need to be resolved. The project builds, but there are lots of type issues that appear in the editor.

Comment (benasher44) [reply .eslintrc.js:46]: If you attempt this after merging this, see ProseMirror/prosemirror#1305, where we discussed issues here in depth. In order to build with NodeNext, here's what I did to start:

1. Update main tsconfig.json moduleResolution to "NodeNext"
2. Override moduleResolution in tests/tsconfig.json to be "node", so cypress tests still build/run
3. Update all prosemirror packages to get the compatible dependencies

Comment (benasher44) [reply .eslintrc.js:46]: @bdbch is auto-fixability a blocker?

Comment (bdbch) [reply .eslintrc.js:46]: Not really. I just need another review from @svenadlung since we share PR responsibilities.

**Review by svenadlung — APPROVED** [review]

**Comment by svenadlung** [comment]
Sorry for the delay on my part. Let's give it a try. Thank you for the contribution @benasher44  🙌

**Comment by benasher44** [comment]
Amazing, thank you!

**Comment by benasher44** [comment]
I did a quick pass on a handful of open PRs in the last week and left comments, but expect to maybe have to quickly follow up with fixes on develop, if people aren't up-to-date with develop

File: .eslintrc.js
Comment (benasher44) [reply .eslintrc.js:46]: Heads up that there is now the `typescript.preferences.importModuleSpecifierEnding` for VSCode, which will make .js endings the default when auto-importing!

**Pull request merged**
